src/Select/Select.tsx

```diff
@@ -1,115 +1,108 @@
-import React, { useState, cloneElement, ReactElement, useEffect, Ref } from 'react'
+import React, { useState, useEffect } from 'react'
 import styled from 'styled-components'
 
 import { BlockProps } from '../Block'
 import { Dropdown, DropdownPlacement } from '../Dropdown'
-import { InputProps } from '../Input/Input'
-import { SelectInput } from '../SelectInput'
+import { SelectInput, SelectInputProps } from '../SelectInput'
 
 import { SelectListProps, SelectListData, SelectList } from '../SelectList'
 
-interface InjectValueAndHandler {
-  onClick: () => void
-  name?: string
-  value?: string
-  placeholder?: string
-  ref?: React.Ref<HTMLInputElement>
-}
+const OptionsWrapper = styled.div<{ fullWidth: boolean }>`
+  width: ${({ fullWidth }) => fullWidth ? '100%' : 'auto'}
+`
 
-interface ChildProps extends Pick<SelectListProps, 'value' | 'data'> {
+interface ListData<D extends SelectListData> extends Pick<SelectListProps<D>, 'value' | 'data'> {}
+
+type InputComponentProps = Partial<Omit<SelectInputProps, 'ref'>>
+
+export interface OptionsProps<D extends SelectListData> extends ListData<D> {
   close: () => void
+  select: (item: D) => void
 }
 
-export interface SelectProps extends Pick<SelectListProps, 'value' | 'data'> {
+export interface SelectProps<D extends SelectListData, I> extends ListData<D> {
   name?: string
   placeholder?: string
-  triggerComponent?: ReactElement<any>
-  onOutsideClick?: (isOpen: boolean, close: Function) => void
-  useTriggerComponentWidth?: boolean
-  onChange?: (value: any) => void
-  test?: string
-  children?: (props: ChildProps) => ReactElement
-}
-
-export const injectComponent = (component: ReactElement<any>, props: InjectValueAndHandler) => {
-  return cloneElement(component, {
-    ...props,
-    ...component.props
-  })
-}
-
-const getTriggerComponent = (props: InjectValueAndHandler, ref: Ref<HTMLInputElement>, triggerComponent?: ReactElement<any>) => {
-  if (triggerComponent) return injectComponent(triggerComponent, { ...props, ref })
-  // Does not need onChange because SelectInput only display data
-  return <SelectInput {...props} ref={ref} />
+  disabled?: boolean,
+  onOutsideClick?: (isOpen: boolean, close: () => void) => void
+  useInputComponentWidth?: boolean
+  onChange: (value: any) => void
+  InputComponent?: React.ComponentType<I & { ref: React.Ref<any> }>
```

Comment:
See comments in Dropdown. Since this pattern seems to be used throughout this PR, we should maybe discuss the general approach we want to take in this package for similar use-cases.

cc @Michiel87 @Stefanwullems